utils.py

```diff
@@ -27,3 +35,47 @@ def piece_av_files(piece: MidiPiece) -> dict:
         "pianoroll_path": pianoroll_path,
     }
     return paths
+
+
+def euclidean_distance(out: torch.Tensor, tgt: torch.Tensor):
+    labels = out.argmax(1).to(float)
+    # euclidean distance
+    return torch.dist(labels, tgt.to(float), p=2)
+
+
+def learning_rate_schedule(step: int, model_size: int, factor: float, warmup: int) -> float:
+    # we have to default the step to 1 for LambdaLR function
+    # to avoid zero raising to negative power.
+    if step == 0:
+        step = 1
+    return factor * (model_size ** (-0.5) * min(step ** (-0.5), step * warmup ** (-1.5)))
+
+
+def load_cached_dataset(cfg: DictConfig, split="test") -> BinsToVelocityDataset:
+    n_dstart_bins, n_duration_bins, n_velocity_bins = cfg.bins.split(" ")
+    n_dstart_bins, n_duration_bins, n_velocity_bins = int(n_dstart_bins), int(n_duration_bins), int(n_velocity_bins)
+
+    config_hash = hashlib.sha256()
+    config_string = json.dumps(OmegaConf.to_container(cfg)) + split
+    config_hash.update(config_string.encode())
+    config_hash = config_hash.hexdigest()
+    cache_dir = "tmp/datasets"
+
+    dataset_cache_file = f"{config_hash}.pkl"
+    dataset_cache_path = os.path.join(cache_dir, dataset_cache_file)
+
+    if os.path.exists(dataset_cache_path):
+        file = open(dataset_cache_path, "rb")
+        dataset = pickle.load(file)
+    else:
+        file = open(dataset_cache_path, "wb")
+        dataset = load_dataset(cfg.dataset_name, split=split)
+        test_dataset = BinsToVelocityDataset(
+            dataset=dataset,
+            n_dstart_bins=n_dstart_bins,
+            n_velocity_bins=n_velocity_bins,
+            n_duration_bins=n_duration_bins,
+            sequence_len=cfg.sequence_size,
+        )
+        pickle.dump(test_dataset, file)
```

Comment:
```suggestion
        hf_dataset = load_dataset(cfg.dataset_name, split=split)
        dataset = BinsToVelocityDataset(
            dataset=hf_dataset,
            n_dstart_bins=n_dstart_bins,
            n_velocity_bins=n_velocity_bins,
            n_duration_bins=n_duration_bins,
            sequence_len=cfg.sequence_size,
        )
        pickle.dump(dataset, file)
```

Was returning a wrong object

Reply:
yup :)